src/map/lua/lua_baseentity.cpp

```diff
@@ -5919,69 +5919,157 @@ uint8 CLuaBaseEntity::getSubJob()
 
 void CLuaBaseEntity::changeJob(uint8 newJob)
 {
-    auto* PChar = dynamic_cast<CCharEntity*>(m_PBaseEntity);
-
-    if (!PChar)
+    if (m_PBaseEntity && m_PBaseEntity->objtype == TYPE_PC)
     {
-        ShowWarning("CLuaBaseEntity::changeJob() - m_pBaseEntity is not of type Char.");
-        return;
-    }
+        auto* PChar = dynamic_cast<CCharEntity*>(m_PBaseEntity);
+
+        if (!PChar)
+        {
+            ShowWarning("CLuaBaseEntity::changeJob() - m_pBaseEntity is not of type Char.");
+            return;
+        }
 
-    JOBTYPE prevjob = PChar->GetMJob();
+        JOBTYPE prevjob = PChar->GetMJob();
 
-    PChar->resetPetZoningInfo();
+        PChar->resetPetZoningInfo();
 
-    PChar->jobs.unlocked |= (1 << newJob);
-    PChar->SetMJob(newJob);
+        PChar->jobs.unlocked |= (1 << newJob);
+        PChar->SetMJob(newJob);
 
-    if (newJob == JOB_BLU)
-    {
-        if (prevjob != JOB_BLU)
+        if (newJob == JOB_BLU)
         {
-            blueutils::LoadSetSpells(PChar);
+            if (prevjob != JOB_BLU)
+            {
+                blueutils::LoadSetSpells(PChar);
+            }
         }
+        else if (PChar->GetSJob() != JOB_BLU)
+        {
+            blueutils::UnequipAllBlueSpells(PChar);
+        }
+        puppetutils::LoadAutomaton(PChar);
+        charutils::SetStyleLock(PChar, false);
+        luautils::CheckForGearSet(PChar); // check for gear set on gear change
+        jobpointutils::RefreshGiftMods(PChar);
+        charutils::BuildingCharSkillsTable(PChar);
+        charutils::CalculateStats(PChar);
+        charutils::CheckValidEquipment(PChar);
+        PChar->PRecastContainer->ChangeJob();
+        charutils::BuildingCharAbilityTable(PChar);
+        charutils::BuildingCharTraitsTable(PChar);
+
+        // clang-format off
+        PChar->ForParty([](CBattleEntity* PMember)
+        {
+            ((CCharEntity*)PMember)->PLatentEffectContainer->CheckLatentsPartyJobs();
+        });
+        // clang-format on
+
+        PChar->UpdateHealth();
+        PChar->health.hp = PChar->GetMaxHP();
+        PChar->health.mp = PChar->GetMaxMP();
+
+        charutils::SaveCharStats(PChar);
+        charutils::SaveCharJob(PChar, PChar->GetMJob());
+        charutils::SaveCharExp(PChar, PChar->GetMJob());
+        PChar->updatemask |= UPDATE_HP;
+
+        PChar->pushPacket(new CCharJobsPacket(PChar));
+        PChar->pushPacket(new CCharStatsPacket(PChar));
+        PChar->pushPacket(new CCharSkillsPacket(PChar));
+        PChar->pushPacket(new CCharRecastPacket(PChar));
+        PChar->pushPacket(new CCharAbilitiesPacket(PChar));
+        PChar->pushPacket(new CCharUpdatePacket(PChar));
+        PChar->pushPacket(new CMenuMeritPacket(PChar));
+        PChar->pushPacket(new CMonipulatorPacket1(PChar));
+        PChar->pushPacket(new CMonipulatorPacket2(PChar));
+        PChar->pushPacket(new CCharSyncPacket(PChar));
```

Comment:
Presumably these big chunks of logic are replicated from somewhere, so it makes sense to extract them into a `charutils` function. Same with the chunk below into `mobutils`

Reply:
please bear with me as I'm learning. To be clear im just adding to the existing binding, so it isn't exactly duplicative, right? is the recommendation to move this entire logic into charutils?

Reply:
Don't worry about it. Just do the recommended styling changes.

Reply:
I think that's everything. Please let me know if I failed to make any of the changes as requested.